lib/acl.c

```diff
@@ -45,12 +45,31 @@ enum {
 
 /* Callback function for when there's no classification match. */
 static inline int
-drop_ipv6_acl_pkts(struct rte_mbuf **pkts, unsigned int num_pkts,
-	__attribute__((unused)) struct gatekeeper_if *iface)
+match_ipv6_acl_pkts(struct rte_mbuf **pkts, unsigned int num_pkts,
+	struct gatekeeper_if *iface)
 {
 	unsigned int i;
-	for (i = 0; i < num_pkts; i++)
-		rte_pktmbuf_free(pkts[i]);
+	for (i = 0; i < num_pkts; i++) {
+		unsigned int j;
+		bool free_pkt = false;
+		/* Start from the first one. */
+		for (j = 1; j < iface->acl_func_count; j++) {
+			int ret = iface->ext_funcs[j](pkts[i], iface);
+			if (ret == -ENOENT)
+				free_pkt = true;
+			else {
```

Comment:
I see your point, but now you uncovered something slightly weird in the code: success (i.e. `ret == 0`) and most errors (i.e. all but `-ENOENT`) lead to `rte_pktmbuf_free()`. Moreover, the match functions do more than matching, it matches and, if successful, submit packets! This is not just a readability issue, this is a problem waiting for someone to have the need to match a packet without submitting it.

There's an easy way to eliminate this unnecessary twister in the code: matching functions just return 0 if there's a match. When there's a match, here we just call `iface->acl_funcs[j](&pkts[i], 1, iface);`. This solution would better match the comment on field `ext_funcs` of `struct gatekeeper_if `, which doesn't mention the importance given to `-ENOENT in the current code.

Finally, submit functions are meant to handle as many packets as possible at once. The current code and the suggested change above don't take advantage of this fact. After the change above is made, the way to send multiple packets to submit functions is to change `process_ipv6_acl()` such that at the step that packets are rearranged in the matrix `pkts`, it also classifies the unmached packets using the matching functions.